src/Data/Machine/MealyM.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE ApplicativeDo #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE InstanceSigs #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Machine.Mealy
+-- License     :  BSD-style (see the file LICENSE)
+--
+-- <http://en.wikipedia.org/wiki/Mealy_machine>
+-- <https://github.com/ivanperez-keera/dunai/blob/develop/src/Data/MonadicStreamFunction/Core.hs#L35>
+-- <https://hackage.haskell.org/package/auto-0.4.3.0/docs/Control-Auto.html>
+-- <https://hackage.haskell.org/package/varying-0.6.0.0/docs/Control-Varying-Core.html>
+----------------------------------------------------------------------------
+module Data.Machine.MealyM
+  ( MealyM(..)
+  , arrPure
+  , arrM
+  , upgrade
+  , scanMealyM
+  , scanMealyMM
+  , embedMealyM
+  , autoMealyM
+  ) where
+
+import Data.Machine
+import Control.Arrow
+import Data.Pointed
+import qualified Control.Category as C
+import Control.Monad.Trans
+import Control.Monad.Identity
+
+-- | 'Mealy' machine, with monadic effects
+newtype MealyM m a b = MealyM { runMealyM :: a -> m (b, MealyM m a b) }
+
+instance Functor m => Functor (MealyM m a) where
+  {-# INLINE fmap #-}
+  fmap f (MealyM m) = MealyM $ \a ->
+    fmap (\(x,y) -> (f x, fmap f y)) (m a)
+
+instance Pointed m => Pointed (MealyM m a) where
+  {-# INLINE point #-}
+  point b = r where r = MealyM (const (point (b, r)))
+
+instance Applicative m => Applicative (MealyM m a) where
+  {-# INLINE pure #-}
+  pure b = r where r = MealyM (const (pure (b, r))) -- Stolen from Pointed
+  MealyM m <*> MealyM n = MealyM $ \a ->
+    do let (ma, na) = (m a, n a)
+       (mb, mm) <- ma
+       (nb, nm) <- na
+       return (mb nb, mm <*> nm)
+
+instance Monad m => Monad (MealyM m a) where
+  {-# INLINE return #-}
+  return = pure
+  (>>=) :: MealyM m a b -> (b -> MealyM m a c) -> MealyM m a c
```

Comment:
Oops. That should allow the removal of the `ScopedTypeVariables` extension if I get rid of this too.